README.md
# flutter_hud_view

一个高可定制化的、无侵入的、支持链式调用的指示条组件（支持进度条指示，自定义提示等）

## Demo

## 使用

在 `pubspec.yml` 文件中添加依赖

```yaml
dependencies:
  flutter_hud_view: ^1.0.0
```

在需要使用的地方import package

```dart
import 'package:flutter_hud_view/flutter_hud_view.dart';
```

基础使用

```dart
/// 初始化
HUDView.of(context).init();

/// 更新部分属性
HUDView.of(context).update();

/// 展示
HUDView.of(context).show();

/// 关闭
HUDView.of(context).dismiss(); /// 或 HUDView.of(context).destroy(); 具体区别请看代码注释
```

## 主要函数

### init()

声明
```dart
init({
  HUDType hudType = HUDType.ANDROID,
  HUDIndicatorType hudIndicatorType = HUDIndicatorType.INDICATOR,
  String text = '',
  Color hudColor,
  Color hudTextColor,
  Color hudBackgroundColor,
  Color backgroundColor,
  Widget customWidget,
  Widget customIndicatorWidget,
  bool tapClose = false,
  bool timerClose = false,
  int timerSeconds,
});
```

参数说明

|参数名		|描述	|默认值   |
|:---------------|:---------------|:-------|
|hudType| HUD类型，HUDType.IOS, HUDType.ANDROID, HUDType.CUSTOM可选|HUDType.ANDROID|
|hudIndicatorType| 指示样式，HUDIndicatorType.INDICATOR(指示器样式，与HUDType相关), HUDIndicatorType.SUCCESS, HUDIndicatorType.FAIL, HUDIndicatorType.CUSTOM可选| HUDIndicatorType.INDICATOR|
|text| 提示文字，不设置不显示|
|hudColor| 指示颜色| HUDType.IOS: Colors.white <br> HUDType.ANDROID: Theme.of(context).accentColor|
|hudTextColor| 提示文字颜色|HUDType.IOS: Colors.white <br> HUDType.ANDROID: Theme.of(context).accentColor|
|hudBackgroundColor| HUDView背景色|HUDType.IOS: Colors.black.withAlpha(200) <br> HUDType.ANDROID: Colors.white|
|backgroundColor|全屏背景色 |HUDType.IOS: Color(0x00FFFFFF) <br> Other: Color(0xA0000000)|
|customWidget| 自定义Widget，当hudType为HUDType.CUSTOM时必填 | null|
|customIndicatorWidget| 指示器Widget，当hudIndicatorType为HUDIndicatorType.CUSTOM时必填 | null|
|tapClose| 点击HUDView时是否关闭 | false|
|timerClose| 是否开启倒计时关闭 | false|
|timerSeconds| 倒计时关闭秒数，timerClose为true时必填| null |


### update()

声明
```dart
update({
  HUDType hudType,
  HUDIndicatorType hudIndicatorType,
  String text,
  Color hudColor,
  Color hudTextColor,
  Color hudBackgroundColor,
  Color backgroundColor,
  Widget customWidget,
  Widget customIndicatorWidget,
  bool tapClose,
  bool timerClose,
  int timerSeconds,
});
```

参数说明

|参数名		|描述	|默认值   |
|:---------------|:---------------|:-------|
|hudType| HUD类型，HUDType.IOS, HUDType.ANDROID, HUDType.CUSTOM可选|HUDType.ANDROID|
|hudIndicatorType| 指示样式，HUDIndicatorType.INDICATOR(指示器样式，与HUDType相关), HUDIndicatorType.SUCCESS, HUDIndicatorType.FAIL, HUDIndicatorType.CUSTOM可选| HUDIndicatorType.INDICATOR|
|text| 提示文字，不设置不显示|
|hudColor| 指示颜色| HUDType.IOS: Colors.white <br> HUDType.ANDROID: Theme.of(context).accentColor|
|hudTextColor| 提示文字颜色|HUDType.IOS: Colors.white <br> HUDType.ANDROID: Theme.of(context).accentColor|
|hudBackgroundColor| HUDView背景色|HUDType.IOS: Colors.black.withAlpha(200) <br> HUDType.ANDROID: Colors.white|
|backgroundColor|全屏背景色 |HUDType.IOS: Color(0x00FFFFFF) <br> Other: Color(0xA0000000)|
|customWidget| 自定义Widget，当hudType为HUDType.CUSTOM时必填 | null|
|customIndicatorWidget| 指示器Widget，当hudIndicatorType为HUDIndicatorType.CUSTOM时必填 | null|
|tapClose| 点击HUDView时是否关闭 | false|
|timerClose| 是否开启倒计时关闭 | false|
|timerSeconds| 倒计时关闭秒数，timerClose为true时必填| null |

### show()

声明
```dart
show({
  autoDestroy = false
});
```

参数说明

|参数名		|描述	|默认值   |
|:---------------|:---------------|:-------|
|autoDestroy| 在HUDView关闭后是否自动销毁，只有tapClose或timerClose为true时有效|false|

### dismiss()

声明
```dart
dismiss();
```

### destroy()

声明
```dart
destroy();
```

## 需要注意的地方

- update()需要在init()后调用
- 调用destroy()后，如需再次显示需要init()

### init和update的使用场景

如果完全不使用update是可以的

使用update可以在同一个context下显示HUDView时复用widget，提高性能

最常见的情况是在联网请求时，当点击按钮后会弹出Loading的HUDView，直到请求结束拿到返回值需要告知用户操作结果

例子代码
```dart
Future request() async {
  HUDView.of(context).init(text: 'Loading').show();
  doRequest(onResponse: () {
    HUDView.of(context).update(hudIndicatorType: HUDIndicatorType.SUCCESS, text: 'Success', timerClose: true, timerSeconds: 2).show(autoDestroy: true);
  });
}
```

如果使用了init，则会覆盖之前所有的update修改

### 为什么要有destroy

由于HUDView针对context的widget进行了缓存，如果不销毁已经失效的context的widget的话，将会出现内存泄漏的情况，所以引入了destroy

那么destroy需要在什么时候使用呢？

- 一次完整的操作结束后
- 页面pop或push之前

例如上面的代码，一次request会展示两次HUDView，所以可以视为一次完整的操作，我们只需在最后一次操作的show函数中设置autoDestroy参数为true即可

或者是当某个页面不再使用了的时候

例如登录，如果登录成功，将会跳转到另一个页面，永远不会再退回来
```dart
Future login() async {
  HUDView.of(context).init(text: 'Loading').show();
  bool result = await login();
  if (result) {
    HUDView.of(context).destroy(); /// 此时使用dismiss()也可以关掉HUDView，但是由于页面会跳转再也不会回来，所以内部的缓存就永久留下了
    Navigator.of(context).push();
  } else {
    HUDView.of(context).update(hudIndicatorType: HUDIndicatorType.FAIL, text: 'Some Error', timerClose: true, timerSeconds: 2).show(autoDestroy: true);  /// 完成了一次完整操作，但是由于后面还会进行再次登录，所以show函数中的autoDestroy就是可选的
  }
}
```

## 开发计划

- 增加带进度的指示器类型

CHANGELOG.md
## [1.0.0] - 2019.09.27

* First version.


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_hud_view/flutter_hud_view.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter HUDView Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FlatButton(
              onPressed: () {
                HUDView.of(context).init(timerClose: true, timerSeconds: 2).show(autoDestroy: true);
              },
              child: Text(
                'Simple Example'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudType: HUDType.IOS,
                  text: 'Loading...',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'iOS Style'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudType: HUDType.ANDROID,
                  text: 'Loading...',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'Android Style'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudIndicatorType: HUDIndicatorType.SUCCESS,
                  text: 'Success',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'Success'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudIndicatorType: HUDIndicatorType.FAIL,
                  text: 'Failure',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'Failure'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudIndicatorType: HUDIndicatorType.CUSTOM,
                  customIndicatorWidget: Icon(
                    Icons.error,
                    color: Theme.of(context).accentColor,
                    size: MediaQuery.of(context).size.width * 0.15,
                  ),
                  text: 'Custom Indicator',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'Custom Indicator Widget'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudType: HUDType.CUSTOM,
                  customWidget: Card(
                    //color: Colors.red,
                    child: Padding(
                      padding: const EdgeInsets.all(20.0),
                      child: Text(
                        'Custom Widget'
                      ),
                    ),
                  ),
                  backgroundColor: Color(0xA0000000),
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                'Custom Widget'
              )
            ),
            FlatButton(
              onPressed: () {
                HUDView.of(context).init(hudType: HUDType.IOS, timerClose: true, timerSeconds: 3).show();
                Timer(Duration(seconds: 2), (){
                  HUDView.of(context).update(hudType: HUDType.ANDROID, text: 'update').show(autoDestroy: true);
                });
              },
              child: Text(
                  'Init and Update'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudType: HUDType.CUSTOM,
                  customWidget: Card(
                    //color: Colors.red,
                    child: Padding(
                      padding: const EdgeInsets.all(20.0),
                      child: Text(
                          'Tap anywhere to close'
                      ),
                    ),
                  ),
                  backgroundColor: Color(0xA0000000),
                  tapClose: true
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                  'Tap close'
              )
            ),
            FlatButton(
              onPressed: () {
                var hud = HUDView.of(context);
                hud.init(
                  hudType: HUDType.IOS,
                  hudColor: Colors.black,
                  hudTextColor: Colors.black,
                  hudBackgroundColor: Colors.white,
                  backgroundColor: Color(0xA0000000),
                  text: 'Custom Color',
                  timerClose: true,
                  timerSeconds: 2
                );
                hud.show(autoDestroy: true);
              },
              child: Text(
                  'Custom Color'
              )
            ),
            FlatButton(
              onPressed: () {
                HUDView.of(context).init(text: '1').show();
                Timer(Duration(seconds: 2), () {
                  HUDView.of(context).dismiss();
                  HUDView.of(context).update(text: '2').show();
                  Timer(Duration(seconds: 2), () {
                    HUDView.of(context).destroy();
                    try {
                      HUDView.of(context).update(text: '3');
                    } catch(e) {
                      /// Can't call update after destroy
                      /// Please call init after destroy
                      HUDView.of(context).init(text: '4', timerClose: true, timerSeconds: 2).show(autoDestroy: true);
                    }
                  });
                });
              },
              child: Text(
                'Dismiss and Destroy'
              )
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/flutter_hud_view.dart
library flutter_hud_view;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'dart:async';

/// [HUDView]为对外开放的唯一接口
/// 通过[HUDView.of]拿到当前context的[_HUD]实例进行相关操作
/// 最简单的调用例子（显示圆圈，点击关闭）
/// HUDView.of(context).init(tapClose: true).show(autoDestroy:true);
class HUDView {

  /// [_hudMap] 保存[_HUD]实例
  /// 建议使用[_HUD].show(autoDestroy=true) 或在页面切换时调用 [_HUD].destroy()
  /// 来销毁实例，以免造成内存泄漏
  static Map<BuildContext, _HUD> _hudMap = {};

  /// 通过[of]从[_hudMap]中获取相关实例
  /// 当[_hudMap]中没有实例的时候会新建一个但不会保存
  /// 只有再调用[_HUD.init()]后才会保存
  static _HUD of(BuildContext context) {
    if (_hudMap.containsKey(context)) {
      return _hudMap[context];
    }
    _HUD hud = _HUD(context);
    return hud;
  }
}


class _HUD {

  /// 调用HUD的[BuildContext]
  BuildContext context;

  /// HUD类型，默认为[HUDType.ANDROID]
  /// [HUDType.IOS] 是默认的iOS样式，透明背景，黑色半透明对话框，白色文字
  /// [HUDType.ANDROID] 是默认的Android样式，灰色半透明背景，白色对话框，
  /// 文字颜色默认为[ThemeData.accentColor]
  /// [HUDType.CUSTOM] 是自定义样式，透明背景，如果设置为[HUDType.CUSTOM],
  /// 则[customWidget]不能为空
  HUDType hudType;

  /// HUD指示类型，默认为[HUDIndicatorType.INDICATOR]
  /// [HUDIndicatorType.INDICATOR] 是默认Loading样式，HUDType相关
  /// [HUDIndicatorType.SUCCESS] 显示Success Icon
  /// [HUDIndicatorType.FAIL] 显示Failure Icon
  /// [HUDIndicatorType.CUSTOM] 可以自定义显示的Icon
  HUDIndicatorType hudIndicatorType;

  /// 提示文字，不设置不显示
  String text;

  /// HUD指示颜色，不设置参考[HUDType]说明
  Color hudColor;

  /// HUD文字颜色，不设置参考[HUDType]说明
  Color hudTextColor;

  /// HUD对话框背景颜色，不设置参考[HUDType]说明
  Color hudBackgroundColor;

  /// 全屏背景颜色，不设置参考[HUDType]说明
  Color backgroundColor;

  /// 自定义显示Widget，当[HUDType]设置为[HUDType.CUSTOM]时有效
  Widget customWidget;

  /// 自定义显示Icon，当[HUDIndicatorType]设置为[HUDIndicatorType.CUSTOM]时有效
  Widget customIndicatorWidget;

  /// 点击对话框外部是否关闭
  bool tapClose;

  /// 是否设置倒计时自动关闭
  bool timerClose;

  /// 倒计时关闭，单位 秒
  int timerSeconds;

  /// 显示状态
  bool _visible;

  /// 自动销毁
  bool _autoDestroy;

  /// 持有的[_HUDWidget.key]
  GlobalKey<_HUDWidgetState> _hudWidgetStateKey;

  /// 倒计时器
  Timer _timer;

  _HUD(this.context) {
    this._hudWidgetStateKey = GlobalKey();
    this._visible = false;
    this._autoDestroy = false;
  }


  /// 初始化函数
  /// 为空的项目将会被赋默认值
  /// 该实例会被保存到[HUDView._hudMap]中
  _HUD init({
    HUDType hudType = HUDType.ANDROID,
    HUDIndicatorType hudIndicatorType = HUDIndicatorType.INDICATOR,
    String text = '',
    Color hudColor,
    Color hudTextColor,
    Color hudBackgroundColor,
    Color backgroundColor,
    Widget customWidget,
    Widget customIndicatorWidget,
    bool tapClose = false,
    bool timerClose = false,
    int timerSeconds,
  }) {
    this.hudType = hudType;
    this.hudIndicatorType = hudIndicatorType;
    this.text = text;
    this.hudColor = hudColor;
    this.hudTextColor = hudTextColor;
    this.hudBackgroundColor = hudBackgroundColor;
    this.backgroundColor = backgroundColor;
    this.customWidget = customWidget;
    this.customIndicatorWidget = customIndicatorWidget;
    this.tapClose = tapClose;
    this.timerClose = timerClose;
    this.timerSeconds = timerSeconds;
    if (this.hudType == HUDType.CUSTOM) assert(customWidget != null);
    if (this.hudIndicatorType == HUDIndicatorType.CUSTOM) assert(customIndicatorWidget != null);
    if (this.timerClose) assert(timerSeconds != null);
    if (this.backgroundColor == null) {
      if (this.hudType == HUDType.ANDROID) {
        this.backgroundColor = Color(0xA0000000);
      } else {
        this.backgroundColor = Color(0x00FFFFFF);
      }
    }
    HUDView._hudMap[context] = this;
    return this;
  }


  /// 更新属性
  /// 如果某项不赋值则保持原来的值
  /// 例如执行
  /// [HUDView].of(context).init(hudType: [HUDType.IOS], text: 'init').update(text: 'update')
  /// 则该HUD的属性为 hudType: [HUDType.IOS], text: 'update'
  _HUD update({
    HUDType hudType,
    HUDIndicatorType hudIndicatorType,
    String text,
    Color hudColor,
    Color hudTextColor,
    Color hudBackgroundColor,
    Color backgroundColor,
    Widget customWidget,
    Widget customIndicatorWidget,
    bool tapClose,
    bool timerClose,
    int timerSeconds,
  }) {
    assert(HUDView._hudMap.containsKey(this.context));
    if (hudType != null) this.hudType = hudType;
    if (hudIndicatorType != null) this.hudIndicatorType = hudIndicatorType;
    if (text != null) this.text = text;
    if (hudColor != null) this.hudColor = hudColor;
    if (hudTextColor != null) this.hudTextColor = hudTextColor;
    if (hudBackgroundColor != null) this.hudBackgroundColor = hudBackgroundColor;
    if (backgroundColor != null) this.backgroundColor = backgroundColor;
    if (customWidget != null) this.customWidget = customWidget;
    if (customIndicatorWidget != null) this.customIndicatorWidget = customIndicatorWidget;
    if (tapClose != null ) this.tapClose = tapClose;
    if (timerClose != null) this.timerClose = timerClose;
    if (timerSeconds != null) this.timerSeconds = timerSeconds;
    if (this.hudType == HUDType.CUSTOM) assert(customWidget != null);
    if (this.hudIndicatorType == HUDIndicatorType.CUSTOM) assert(customIndicatorWidget != null);
    if (this.timerClose) {
      assert(this.timerSeconds != null);
      this._timer?.cancel();
    }
    return this;
  }


  /// 展示HUD
  /// [autoDestroy] 自动销毁，仅在[timerClose] = true时有效
  /// 关闭时自动销毁该[context]在[HUDView]中的实例
  _HUD show({autoDestroy = false}) {
    _autoDestroy = autoDestroy;
    if (timerClose) {
      _timer?.cancel();
      _timer = Timer(Duration(seconds: timerSeconds), () {
        if (_autoDestroy) {
          this.destroy();
        } else {
          this.dismiss();
        }
      });
    }
    if (!_visible) {
      Navigator.of(context).push(
          PageRouteBuilder(
              opaque: false,
              pageBuilder: (context, _, __) {
                _visible = true;
                return _HUDWidget(_hudWidgetStateKey, this);
              }));
    } else {
      _hudWidgetStateKey.currentState.updateConfig(this);
    }
    return this;
  }

  /// 关闭HUD
  /// 如果在关闭后还需要调用[update]，请使用此函数
  /// 如果在关闭后不再需要[update]，请使用[destroy]
  _HUD dismiss() {
    this._timer?.cancel();
    _autoDestroy = false;
    if (_visible) {
      Navigator.of(_hudWidgetStateKey.currentContext).pop();
      _hudWidgetStateKey = GlobalKey();
      _visible = false;
    }
    return this;
  }


  /// 销毁HUD
  /// 在关闭该HUD的基础上销毁了该[context]在[HUDView]中的实例
  /// 销毁后再调用[update]会报错，如需使用请重新调用[init]
  _HUD destroy() {
    this.dismiss();
    HUDView._hudMap.remove(context);
    return this;
  }
}

enum HUDType {
  ANDROID,
  IOS,
  CUSTOM,
}

enum HUDIndicatorType {
  INDICATOR,
  SUCCESS,
  FAIL,
  CUSTOM,
}

class _HUDWidget extends StatefulWidget {
  final _HUD config;

  _HUDWidget(Key key, this.config) : super(key:key);

  @override
  _HUDWidgetState createState() => _HUDWidgetState();
}

class _HUDWidgetState extends State<_HUDWidget> {
  _HUD _config;

  void updateConfig(_HUD newConfig) {
    setState(() {
      _config = newConfig;
    });
  }

  @override
  void initState() {
    super.initState();
    _config = widget.config;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: _config.backgroundColor,
      body: GestureDetector(
        child: Container(
            color: Color(0x00FFFFFF),
            constraints: BoxConstraints.expand(),
            child: _getCenterContent()
        ),
        onTap: () {
          if (_config.tapClose) {
            if (_config._autoDestroy) {
              _config.destroy();
            } else {
              _config.dismiss();
            }
          }
        },
      ),
    );
  }

  Widget _getCenterContent() {
    return Center(
      child: _config.hudType == HUDType.CUSTOM ? _config.customWidget : Container(
        height: MediaQuery.of(context).size.width * 0.3,
        padding: EdgeInsets.all(10.0),
        decoration: BoxDecoration(
          color: _getHUDBackgroundColor(),
          borderRadius: BorderRadius.circular(10.0),
        ),
        constraints: BoxConstraints(
          minWidth: MediaQuery.of(context).size.width * 0.3,
          maxWidth: MediaQuery.of(context).size.width * 0.8,
        ),
        child: Column(
          //mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            _getIndicatorWidget(),
            Offstage(
              offstage: _config.text.isEmpty,
              child: SizedBox(height: _config.hudIndicatorType == HUDIndicatorType.INDICATOR ? 20.0 : 10.0,),
            ),
            Offstage(
              offstage: _config.text.isEmpty,
              child: Text(
                _config.text,
                style: TextStyle(
                    color: _getHUDTextColor(),
                    fontWeight: FontWeight.w500
                ),
              ),
            )
          ],
        ),
      ),
    );
  }

  Color _getHUDBackgroundColor() {
    if (_config.hudBackgroundColor != null) return _config.hudBackgroundColor;
    if (_config.hudType == HUDType.ANDROID) return Colors.white;
    if (_config.hudType == HUDType.IOS) return Colors.black.withAlpha(200);
    return Color(0x00FFFFFF);
  }

  Color _getHUDColor() {
    if (_config.hudColor != null) return _config.hudColor;
    if (_config.hudType == HUDType.ANDROID) return Theme.of(context).accentColor;
    if (_config.hudType == HUDType.IOS) return Colors.white;
    return Color(0x00FFFFFF);
  }

  Color _getHUDTextColor() {
    if (_config.hudTextColor != null) return _config.hudTextColor;
    if (_config.hudType == HUDType.ANDROID) return Theme.of(context).accentColor;
    if (_config.hudType == HUDType.IOS) return Colors.white;
    return Color(0x00FFFFFF);
  }

  Widget _getIndicatorWidget() {
    if (_config.hudIndicatorType == HUDIndicatorType.CUSTOM) return _config.customIndicatorWidget;
    if (_config.hudIndicatorType == HUDIndicatorType.SUCCESS) return Icon(Icons.check, color: _getHUDColor(), size: MediaQuery.of(context).size.width * 0.15,);
    if (_config.hudIndicatorType == HUDIndicatorType.FAIL) return Icon(CupertinoIcons.clear_thick, color: _getHUDColor(), size: MediaQuery.of(context).size.width * 0.15,);
    if (_config.hudType == HUDType.ANDROID) return CircularProgressIndicator(
        valueColor: new AlwaysStoppedAnimation(_getHUDColor()));
    if (_config.hudType == HUDType.IOS) return CupertinoActivityIndicator(
      radius: MediaQuery.of(context).size.width * 0.05,
    );
    return Container();
  }
}
